tokio-util/src/task/spawn_pinned.rs

```diff
@@ -0,0 +1,184 @@
+use std::any::Any;
+use std::fmt;
+use std::fmt::{Debug, Formatter};
+use std::future::Future;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::sync::mpsc::{channel, Sender};
+use std::sync::Arc;
+use tokio::runtime::Builder;
+use tokio::sync::mpsc::{unbounded_channel, UnboundedReceiver, UnboundedSender};
+use tokio::task::{spawn_blocking, spawn_local, JoinHandle, LocalSet};
+
+/// Create a new pool of threads to handle `!Send` tasks. Spawn tasks onto this
+/// pool via [`LocalPoolHandle::spawn_pinned`].
+pub fn new_local_pool(pool_size: usize) -> LocalPoolHandle {
+    assert!(pool_size > 0);
+
+    let workers = (0..pool_size)
+        .map(|_| LocalWorkerHandle::new_worker())
+        .collect();
+
+    let pool = Arc::new(LocalPool { workers });
+
+    LocalPoolHandle { pool }
+}
+
+/// A handle to a local pool created by [`new_local_pool`]
+#[derive(Clone)]
+pub struct LocalPoolHandle {
+    pool: Arc<LocalPool>,
+}
+
+impl LocalPoolHandle {
+    /// Spawn a task onto a worker thread and pin it there so it can't be moved
+    /// off of the thread. Note that the future is not Send, but the `FnOnce` which
+    /// creates it is.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use std::rc::Rc;
+    /// use tokio_util::task::new_local_pool;
+    ///
+    /// #[tokio::main]
+    /// async fn main() {
+    ///     // Create the local pool
+    ///     let pool = new_local_pool(1);
+    ///
+    ///     // Spawn a !Send future onto the pool and await it
+    ///     let output = pool
+    ///         .spawn_pinned(|| {
+    ///             // Rc is !Send + !Sync
+    ///             let local_data = Rc::new("test");
+    ///
+    ///             // This future holds an Rc, so it is !Send
+    ///             async move { local_data.to_string() }
+    ///         })
+    ///         .await
+    ///         .unwrap();
+    ///
+    ///     assert_eq!(output, "test");
+    /// }
+    /// ```
+    pub fn spawn_pinned<Fut: Future + 'static>(
+        &self,
+        create_task: impl FnOnce() -> Fut + Send + 'static,
+    ) -> JoinHandle<Fut::Output>
+    where
+        Fut::Output: Send + 'static,
+    {
+        self.pool.spawn_pinned(create_task)
+    }
+}
+
+impl Debug for LocalPoolHandle {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        f.write_str("LocalPoolHandle")
+    }
+}
+
+struct LocalPool {
+    workers: Vec<LocalWorkerHandle>,
+}
+
+impl LocalPool {
+    /// Spawn a `?Send` future onto a worker
+    fn spawn_pinned<Fut: Future + 'static>(
+        &self,
+        create_task: impl FnOnce() -> Fut + Send + 'static,
+    ) -> JoinHandle<Fut::Output>
+    where
+        Fut::Output: Send + 'static,
```

Comment:
`spawn_pinned` is now async, and `spawn_pinned_blocking` is the PR's original function (blocks when receiving the join handle).

Reply:
Updated by making `spawn_pinned` synchronously send the task but return a future for the join handle (it can be safely dropped if the join handle isn't needed). I think that resolves this comment thread.